compiler/ast.nim

```diff
@@ -1128,7 +1128,17 @@ proc discardSons*(father: PNode)
 type Indexable = PNode | PType
 
 proc len*(n: Indexable): int {.inline.} =
-  result = n.sons.len
+  when n is PNode:
```

Comment:
Instead use `safeLen` where required, don't slow down good old `len` please.

Reply:
I can do that, but note that all checkSons, loops, etc... are now prone to break as they no longer guard on nil like they used to.

I know it's used _everywhere_, something better would be to somehow force the nil check as I believe the variant's branch will still be nil. 🤔 Don't suppose you got a sneak trick to do that? 🥷

Reply:
I don't understand. All I'm asking for is that you don't touch `ast.nim`.

Reply:
If one provides a malformed AST now it will likely crash the compiler, without providing an ill formed AST message.

I can double check the symptom but should be replicable with the test I provided on a version before this change and after the removal of non-nil sequences.

Reply:
Malformed ASTs are not `nil`, they are `nkEmpty` (and hopefully soon `nkError`).

Reply:
I know that and it's too early to even detect it, IIRC we blindly sem the return of a macro, so all of that is not at all going to help. :(

The issue is that semIf presently crashes as the malformed AST is passed in. All the usual guards, including checkMinSons will also crash. This will be pervasive throughout the compiler and not detected by tests because we don't fuzz on macro input.

No matter what there are likely going to be a ton of guards put everywhere now so who knows what that impact will be. We'll just take a different performance hit as the work needs to be fundamentally done. Dintner elsewhere will hopefully involve less tight loops.

I'll pull this part out of the PR but there is likely a very significant regression in the inability to handle malformed macro and other inputs after the removal of no nil seqs. Simply saying use safeLen which itself is buggy (should have the updated case on kind) and will need to be changed. checkMinSons and the like will more than likely have to use safeLen require the introduction of that in many more areas. The better approach is likely putting safeLen in checkSons and friends.

I do agree slowing down len sucks because it's used everywhere including tight loops and this doesn't need to be the approach or is even the right approach to begin with. I do want to ensure I've clarified the pervasive nature of the issue. My hope was that you'd see a more central area or layer to detect/guard against it. Again I'll pull it out here and hopefully a better plan of attack can be sorted out elsewhere.

Reply:
I don't agree but let's this dicuss it elsewhere.